### PR TITLE
fix: publishing modules

### PR DESCRIPTION
**Pull request by bang9**
- Fix runtime error due to publishing modules
- Refactor PubSub

ticket: [CLNP-1469]

[CLNP-1469]: https://sendbird.atlassian.net/browse/CLNP-1469?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by AhyoungRyu — APPROVED** [review]
👍👍👍

File: src/lib/pubSub/topics.ts [hunk 1]
```diff
@@ -1,20 +1,107 @@
...
+    }
+  | {
+      topic: PUBSUB_TOPICS.UPDATE_OPEN_CHANNEL;
+      payload: OpenChannel;
+    };
+
+export type SBUGlobalPubSub = PubSubTypes<
+  PUBSUB_TOPICS,
+  SBUGlobalPubSubTopicPayloadUnion
+>;
```
Comment (liamcho) [inline src/lib/pubSub/topics.ts:106]: nice!

File: src/lib/pubSub/index.ts [hunk 2]
```diff
@@ -1,43 +1,50 @@
...
 }
+  ? P
+  : never;
 
-const pubSubFactory = (): PubSubTypes => {
-  const topics = {};
-  const hOP = topics.hasOwnProperty;
+const pubSubFactory = <
+  T extends keyof any = keyof any,
+  U extends { topic: T; payload: any } = any
```
Comment (liamcho) [inline src/lib/pubSub/index.ts:27]: nice but why we need default values here?
```
= keyof any,
= any
```

File: src/lib/pubSub/index.ts [hunk 3]
```diff
@@ -1,43 +1,50 @@
...
 
-export interface PubSubTypes {
-  __getTopics: () => Record<string, string>;
-  subscribe: (topic: string, listener: unknown) => { remove: () => void };
-  publish: (topic: string, info: unknown) => void;
+export type PubSubTypes<
+  T extends keyof any,
+  U extends { topic: T; payload: any }
+> = {
+  __getTopics: () => Record<string, Set<(data: PayloadByTopic<T, U>) => void>>;
```
Comment (liamcho) [inline src/lib/pubSub/index.ts:10]: should this not be
```
  __getTopics: () => Record<T, Set<(data: PayloadByTopic<T, U>) => void>>;
```
?

File: src/lib/pubSub/index.ts (hunk 2 above)
Comment (bang9) [reply src/lib/pubSub/index.ts:27]: for pubSubFactory testing

File: src/lib/pubSub/index.ts (hunk 3 above)
Comment (bang9) [reply src/lib/pubSub/index.ts]: nice catch

**Review by HoonBaek — APPROVED** [review]

File: src/modules/internalInterfaces.ts [hunk 4]
```diff
@@ -2,3 +2,17 @@ export enum PublishingModuleType {
   CHANNEL = 'CHANNEL',
   THREAD = 'THREAD',
 }
+
+// NOTE: To maintain compatibility with the sendbirdSelector provided to customers
+//  this utility function ensures that if publishingModules is not provided or its length is zero
+//  it will maintain the same behavior as before.
+const publishTo = (target: PublishingModuleType, modules?: PublishingModuleType[]) => {
```
Comment (AhyoungRyu) [inline src/modules/internalInterfaces.ts:9]: nit; this sounds like an action handler which helps to publish modules to somewhere whereas it determines whether the target modules are published or not. 
How about `publishedTo` ?(not sure this what you meant 😅)

**Review by AhyoungRyu — APPROVED** [review]

File: src/modules/internalInterfaces.ts (hunk 4 above)
Comment (liamcho) [reply src/modules/internalInterfaces.ts:9]: Namespace is somewhat confusing.

change to something more explicit:
- isModuleTypeIncludedInModules

**Review by liamcho — APPROVED** [review]
LGTM, Nice work @bang9 !

File: src/modules/internalInterfaces.ts (hunk 4 above)
Comment (bang9) [reply src/modules/internalInterfaces.ts:9]: ㅋㅋㅋ I wrote it without thinking properly. I'll make the necessary corrections.

Comment (bang9) [reply src/modules/internalInterfaces.ts:9]: https://github.com/sendbird/sendbird-uikit-react/pull/842

**Pull request merged**